src/chainstate/stacks/db/blocks.rs

```diff
@@ -4859,6 +4869,135 @@ impl StacksChainState {
+
+        // go from oldest burn header hash to newest
+        for ancestor_bhh in ancestor_burnchain_header_hashes.iter().rev() {
+            let stacking_ops = SortitionDB::get_stack_stx_ops(sortdb_conn, ancestor_bhh)?;
+            let transfer_ops = SortitionDB::get_transfer_stx_ops(sortdb_conn, ancestor_bhh)?;
+
+            for stacking_op in stacking_ops.into_iter() {
+                if processed_burnchain_txids.contains(&stacking_op.txid) {
+                    continue;
+                }
+                all_stacking_burn_ops.push(stacking_op);
+            }
+
+            for transfer_op in transfer_ops.into_iter() {
+                if processed_burnchain_txids.contains(&transfer_op.txid) {
+                    continue;
+                }
+                all_transfer_burn_ops.push(transfer_op);
+            }
+        }
+        Ok((all_stacking_burn_ops, all_transfer_burn_ops))
+    }
+
+    /// Get the list of burnchain-hosted stacking and transfer operations to apply when evaluating
+    /// the Stacks block that was selected for this burnchain block.
+    /// The rules are different for different epochs:
+    ///
+    /// * In Stacks 2.0/2.05, only the operations in the burnchain block will be considered.
+    /// So if a transaction was mined in burnchain block N, it will be processed in the Stacks
+    /// block mined in burnchain block N (if there is one).
+    ///
+    /// * In Stacks 2.1+, the operations in the last K burnchain blocks that have not yet been
+    /// considered in this Stacks block's fork will be processed in the order in which they are
+    /// mined in the burnchain.  So if a transaction was mined in an burnchain block between N and
+    /// N-K inclusive, it will be processed in each Stacks fork that contains at least one Stacks
+    /// block mined in the same burnchain interval.
+    ///
+    /// The rationale for the new behavior in Stacks 2.1+ is that burnchain-hosted STX operations
+    /// can get picked up in Stacks blocks that only live on short-lived forks, or get mined in
+    /// burnchain blocks in which there was no sortiton.  In either case, the operation does not
+    /// materialize on the canonical Stacks chain.  This is a bad user
+    /// experience, because the act of sending a PreStxOp plus this StackStxOp / TransferStxOp is a
+    /// time-consuming and tedious process that must then be repeated.
+    ///
+    /// The change in Stacks 2.1+ makes it so that it's overwhelmingly likely to work
+    /// the first time -- the choice of K is significantly bigger than the length of short-lived
+    /// forks or periods of time with no sortition than have been observed in practice.
+    fn get_stacking_and_transfer_burn_ops<'b>(
+        chainstate_tx: &'b mut ChainstateTx,
+        parent_index_hash: &StacksBlockId,
+        sortdb_conn: &Connection,
+        burn_tip: &BurnchainHeaderHash,
+        burn_tip_height: u64,
+    ) -> Result<(Vec<StackStxOp>, Vec<TransferStxOp>), Error> {
```

Comment:
I don't think the lifetime generic is necessary here: rust can almost always figure out the lifetimes for references that don't land in the return type.